packages/instruments/README.md

```diff
@@ -0,0 +1,41 @@
+## `@envelop/instruments`
+
+This package contains uitility functions and types to ease the use of instruments accross Envelop,
+Yoga, wathwg-node and Hive Gateway plugins.
+
+### `getInstrumentsAndPlugins(plugins: Plugin[]): { pluginInstruments: Instruments[], plugins: Plugin[] }`
+
+This function extracts the instruments from the plugins and returns both the extracted instruments
+and the plugins without their `instruments` field.
+
+This is usefull when you want to customize the execution order of the instruments.
+
+```ts
+import { getInstrumentsAndPlugins } from '@envelop/instruments'
+
+const { pluginInstruments, plugins } = getInstrumentsAndPlugins([
+  // put you plugin list here. This list can contain plugins with and without instruments.
+])
+```
+
+## `composeInstruments(instruments: Instruments[]): Instruments`
+
+This function composes all the instruments into one. The instruments will be called in the same
+order than they are in the array (first is outter most call, last is inner most).
+
+This can be used in conjonction with `getInstrumentsAndPlugins` function to customize the order of
+execution of the instruments if the default one doesn't suites your needs.
+
+```ts
+import { getInstrumentsAndPlugins } from '@envelop/instruments'
+
+const { pluginInstruments, plugins } = getInstrumentsAndPlugins([
+  // put you plugin list here. This list can contain plugins with and without instruments.
+])
+
+const instruments = composeInstruments(pluginInstruments)
+
+const getEnveloped = envelop({
+  plugins: [...plugins, { instruments }]
```

Comment:
the point of this then is to be able to extract the instruments from an existing list of plugins, and then tack them on the end of the plugin list?

This is an interesting workaround and could work. An alternative approach might be to set a priority number on plugins, so that regardless of the order they are defined in, they are executed based on priority. We could then allocate certain ranges for certain things.

Reply:
That's closed to the idea yes.

We already have a defined order of plugins execution : It's the order they are defined in the array.
The problem is that, in the case of "wrapping phases", there is often the need to have a different order for "wrapping" and for normal hooks.

So the point of those utils are to extract instruments from plugins (and remove them), so that you can compose them is the order you want. Since you know there is no other instruments, you can put it in the end of the plugin, or at the beginning, it does matter :-)